### PR TITLE
nix-copy-closure: Implement in C++.

### PR DESCRIPTION
**Pull request by shlevy**
(on top of #981)

### CODE REVIEW

File: src/libstore/store-api.hh [hunk 1]
```diff
@@ -552,7 +552,7 @@ void removeTempRoots();
 ref<Store> openStore(const std::string & uri = getEnv("NIX_REMOTE"));
 
 
-void copyPaths(ref<Store> from, ref<Store> to, const Paths & storePaths);
+void copyPaths(ref<Store> from, ref<Store> to, const Paths & storePaths, bool substitute = false);
```
Comment (shlevy) [inline src/libstore/store-api.hh]: @k0001 FYI if this gets merged it will be easy to implement #937 in the new c++ build-remote

Comment (k0001) [inline src/libstore/store-api.hh]: Thanks for keeping this in mind @shlevy ! I'm really looking forward to having #937 in `master`.

**Comment by shlevy** [comment]
@edolstra What path do you want to go here? Summarizing from previous discussion, I think the two reasonable options are:

1. package `nix-copy-closure.pl` with the `nix-perl` package, install together by default, but with a configure option to build them separately (or, of course, not build the nix-perl stuff at all), recommend new users use `nix copy`
2. Do a blind rewrite of nix-copy-closure.pl and attendant `nix-store --serve` code to C++.

I *strongly* prefer 1, as it will give people who don't want perl a way to get it without breaking compat immediately and requiring a decent chunk of work to port something that will just go away eventually, but I will do either of these.

**Comment by shlevy** [comment]
@edolstra rebased on master. Still need advice re: previous comment.

**Comment by copumpkin** [comment]
👍 for option 2 😄

**Comment by shlevy** [comment]
@edolstra ping

**Comment by shlevy** [comment]
@edolstra I would really like to wrap this up. Can you please provide feedback here? If I haven't heard anything in a week or so I will go ahead with option 1 above, *please* let me know before then if that's not likely to be merged so I don't waste my time.

**Comment by edolstra** [comment]
Thanks, merged.

**Comment by shlevy** [comment]
Thank you!

**Pull request merged**